### PR TITLE
Support comma-delimited multi `@Param` or `@Header`

### PR DESCRIPTION
**Pull request by ghkim3221**
Motivation:

 - Some APIs need comma-delimited multi parameters: `/hello1?number=1,2,3`.
   - https://www.baeldung.com/spring-request-param#mapping-a-multi-value-parameter

Modifications:

- Added `@Delimiter` annotation that splits the value of query parameter or header
  with the specified delimiter
- Added `AnnotatedServiceBindingBuilder.queryDelimiter()` that splits the value of query parameter with the specified delimiter

Result:

- A comma-delimited query parameter or header can be split easily.

### CODE REVIEW

**Comment by KarboniteKream** [comment]
After this is merged, we could remove [this page](https://armeria.dev/docs/advanced-client-interoperability)

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Should we modify `QueryParams.fromQueryString(queryString)` rather than fix it only in annotated service?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:681]: Question: What happen if a client sends `params=1,2,3&params=4,5,6`? Should it be converted to `List(1,2,3,4,5,6)` or `List("1,2,3", "4,5,6")`?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: > Can we modify the decoder?

Sure. 🙂 If you need to modify it, you can do it.
One remaining concern is a breaking change. If we change the default behavior, I guess, it could be an unexpected change for some users.
Do you think this could be the default behavior?
Otherwise, should we add a flag like `Flags.allowCommaDelimitedQueryParameters()` and let users turn on the feature manually? If so, I think we can turn on it by default in Armeria 2.0.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: For the sake of convenience of Spring users, we may add the property to `ArmeriaSettings`

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:681]: As multiple `params`s are used, `List("1,2,3", "4,5,6")` makes sense.

>  Would it be better to make comma-delimited optional?

I'm not sure. Even though it is not a standard protocol, it is widely used for many libraries. I am OK to enable it by default. 
@trustin @jrhee17 Any thought on this?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Because it's not a standard, + 1 for adding Flags.

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:681]: How do others do such cases?

Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:681]: For the following curl:
```
curl -v "http://localhost:8080?params=1,2,3"
curl -v "http://localhost:8080?params=1,2,3&params=4,5,6"
```

### Spring
API:
```
@GetMapping("/")
public ResponseEntity<String> multiParams(@RequestParam("params") List<String> params) {
    return ResponseEntity.ok(String.join(":", params));
}
```
Received:
```
1:2:3
1,2,3:4,5,6
```

Let me check other frameworks soon.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:681]: > Let me check other frameworks soon.

I don't think we don't have to check other frameworks. 😄 
Let's approach in the same direction.

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Because the `QueryParams` doesn't know the target class to resolve, it will always delimit query parameters if contain the comma. That is, we should delimit `@Param List<String> foo` but shouldn't `@Param String foo`. But Armeria will resolve the first element if the corresponding value contains the comma.
So, modifying `QueryParams` isn't suitable I guess.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Good point. It could be vague and cause extra confusion. 
It might be clear to only apply support for comma-delimited to annotated services.
Optionally, if users don't want to turn on the flag (`Flags.allowCommaDelimitedQueryParameters()`) globally, we might provide `@Delimited(",")` you said in https://github.com/line/armeria/pull/4060#discussion_r806616127.

File: site/src/pages/docs/server-annotated-service.mdx
Comment (jrhee17) [inline site/src/pages/docs/server-annotated-service.mdx:308]: 1) The sentence starts with `the type`, but it is unclear what type it is
2) The sentence starting with `Or` is not a complete sentence.

What do you think of the following?

```suggestion
If multiple parameters exist with the same name in a query string, they can be injected as a `List<?>`
or `Set<?>`, e.g. `/hello1?number=1&number=2&number=3`. If a single parameter is mapped to a
`List<?>` or `Set<?>`, it will be converted into comma-separated values, e.g. `/hello1?number=1,2,3`.
You can use <type://@Default> annotation or `Optional<?>` class here, too.
```

**Review by jrhee17 — COMMENTED** [review]
Looks good overall! 👍 🙇  Left some minor comments

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Yeah, the flag looks good to me also 👍 We can introduce a kind of `@Delimited` when we find a better solution.

BTW, how can I test using the flag...? Would it be better to add `allowCommaDelimitedQueryParameters` method that has `Flags.allowCommaDelimitedQueryParameters()` by default to `AnnotatedServiceBindingBuilder`?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: > Would it be better to add allowCommaDelimitedQueryParameters method that has Flags.allowCommaDelimitedQueryParameters() by default to AnnotatedServiceBindingBuilder?

That's a good idea. 👍

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: Our flags were so far mostly for changing some thresholds (timeouts, cache sizes, ..). I agree that a user should be given some options for convenience, but are we sure we want to introduce a flag for that? How about just starting from introducing an additional parameter to `QueryParams.fromQueryString()` and `QueryParams.toQueryString()`? e.g.

```java
p = QueryParams.fromQueryString("a=1;a=2,3", QueryStringDecoderOptions.SEMICOLON_AS_SEPARATOR, QueryStringDecoderOptions.COMMA_SOMETHING_SOMETHING)

assert(
  "a=1,2,3".equals(
    p.toQueryString(QueryStringEncoderOptions.COMMA_SOMETHING_SOMETHING)
  )
)
```

Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: I introduced `useQueryDelimiter` for building an annotated service:

```java
sb.annotatedService()
  .useQueryDelimiter(",")
  .build(new MyAnnotatedService());
```

Actually, we don't have to introduce a new flag because it makes sense that not specifying `useQueryDelimiter` means not allowing delimition.  WDYT?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:203]: Wondering if we want to limit this feature to only direct usage of `@Param`/`@Header` only in the service method.

i.e. The following behavior would be inconsistent
```
static class Simple {
    @Param
    public List<String> l;
}

@RegisterExtension
static final ServerExtension server = new ServerExtension() {
    @Override
    protected void configure(ServerBuilder sb) throws Exception {
        sb.annotatedService()
          .pathPrefix("/")
          .useQueryDelimiter(",")
          .build(new Object() {

            @Get("/requestObject")
            public String requestObject(@RequestObject Simple simple) {
                return Joiner.on(':').join(simple.l);
            }

            @Get("/justParam")
            public String justParam(@Param List<String> l) {
                return Joiner.on(':').join(l);
            }
        });
    }
};

@Test
void test() {
    AggregatedHttpResponse response = server.webClient().prepare().get("/requestObject")
                                            .queryParam("l", "a,b,c")
                                            .execute().aggregate().join();
    System.out.println(response.contentUtf8());

    response = server.webClient().prepare().get("/justParam")
                     .queryParam("l", "a,b,c")
                     .execute().aggregate().join();
    System.out.println(response.contentUtf8());
}
```

**Review by jrhee17 — COMMENTED** [review]
Looking good! Left some minor comments 🙏

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (jrhee17) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:130]: > Actually, we don't have to introduce a new flag because it makes sense that not specifying useQueryDelimiter means not allowing delimition. WDYT?

I can't speak for @trustin on his original intention, but I agree with your statement.
This feature is `annotationService` specific, so I guess it defining as a flag is unnecessary.
I guess just remove the `Flags.java` implementation and by default `queryDelimiter = null` makes sense 👍

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:203]: This is a good point and I feel like a user may need more customization than what we have in this PR. For example, a user may want to use the delimiter only for query string and not for headers, or only for a certain query parameter/header.

For example:

```java
sb.annotatedService()
  .queryParamDelimiter(",") // For all query params
  .queryParamDelimiter("foo", ",") // Only for "foo"
  .queryParamDelimiter(queryParamName -> ",") // dynamically
  .headerDelimiter(...)
  ...
```

Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:203]: At first, I imagined a '@Delimiter' annotation which represents what query delimiter to use. In this case, a user can specify the annotation with the specific '@Param' or '@Header' to enable the query delimitor to the specific param.

'''java
public void foo(@Header @Delimiter(",") List<String> delimited, @Header List<List> notDelimited);
'''

HDYT?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:203]: > ```
> public void foo(@Header @Delimiter(",") List<String> delimited,
>                 @Header List<List> notDelimited);
> ```

Sounds good to me. :+1:

Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: Added `Delimiter` annotation, it has precedence over annotated service settings. And removed the flag.

Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: > Wondering if we want to limit this feature to only direct usage of @Param/@Header only in the service method.

> For example, a user may want to use the delimiter only for query string and not for headers, or only for a certain query parameter/header.

Would it be better to support delimiter for all of `@Param`, `@Header`, and `@RequestObject`? Currently, it's only available for the `@Param`.

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: > Would it be better to support delimiter for all of `@Param`, `@Header`, and `@RequestObject`? Currently, it's only available for the `@Param`.

Yeah, it's be nice if possible. How hard would it be? How would you support it for `@RequestObject` by the way?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: > it has precedence over annotated service settings. And removed the flag.

I still think the flag is valuable. Because some users don't want to specify `@Delimeter` for every method. It may force users to do the same thing repeatedly.
They need to set the default allowed delimiters globally. For example: `-D...allowCommaDelimitedQueryParameters=&,",",;`.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: Otherwise, we can allow to configure the default delimiters with `ServerBuilder` like Trustin said.

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java:1128]: Use Armeria client?

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:134]: Dead code?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:164]:
```suggestion
     * Sets the delimeter for a query parameter value. 
     * Multiple values delimited by the specified by {@code delimiter} will be automatically split into a list of values.
     * It is disabled by default.
```

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:166]: Could we add an example?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:169]: I think we can omit the `use` prefix.
```suggestion
    public AnnotatedServiceBindingBuilder queryDelimiter(String delimiter) {
```

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: > I still think the flag is valuable. Because some users don't want to specify `@Delimeter` for every method.

I agree that there should be a way to specify the delimiter for a wider scope, such as on a per-server basis. However, I'm not sure if it's worth introducing a flag for this.

> Otherwise, we can allow to configure the default delimiters with `ServerBuilder` like Trustin said. 

+1

**Review by trustin — APPROVED** [review]
Excellent work, @ghkim3221 ❤️ LGTM once all other folks' comments are addressed. Thanks a lot for your patience!

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java:177]: Not related with this PR, but we could make it less error prone to add a builder method to `AnnotatedServiceBindingBuilder` a little bit: https://github.com/line/armeria/issues/4205

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
Comment (ghkim3221) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java:1128]: Could you explain more? In the `AnnotatedServiceTest`, almost all of the tests including assertions (`testBody`) are using `CloseableHttpClient`.

Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java:1128]: Please ignore the comment. I guess the tests were added before `WebClient` was made.
We can clean up the legacy code later. That is not related to this PR.

File: site/src/pages/docs/server-annotated-service.mdx
Comment (ikhoon) [inline site/src/pages/docs/server-annotated-service.mdx:327]:
```suggestion
If you specify query delimiter while building an annotated service, or a <type://@Delimiter> annotation with a query
```

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:207]: If we introduce a method to `ServerBuilder`, may users be confused? `Server.builder().queryDelimiter(",")` isn't clear that it looks applicable for both `HttpService` and `AnnotatedService`. Or did you imagine such as `annotatedServiceQueryDelimiter()`?

File: site/src/pages/docs/server-annotated-service.mdx
Comment (ghkim3221) [reply site/src/pages/docs/server-annotated-service.mdx]: Just a question, what is difference between ``Delimiter`` and `<type://@Delimiter>`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:169]: Added! Could you take a look?

File: site/src/pages/docs/server-annotated-service.mdx
Comment (ikhoon) [reply site/src/pages/docs/server-annotated-service.mdx]: Good question. `<type://@Delimiter>` will link to the Javadoc like:
image

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:169]: Looks good!

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:173]:
```suggestion
     *     <li>{@code ?query=a,b,c&query=d,e,f} will be resolved to {@code "a,b,c"} and {@code "d,e,f"}</li>
```

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java:181]:
```suggestion
     *     <li>{@code ?query=a,b,c&query=d,e,f} will be resolved to {@code "a,b,c"} and {@code "d,e,f"}</li>
```

File: site/src/pages/docs/server-annotated-service.mdx
Comment (ikhoon) [inline site/src/pages/docs/server-annotated-service.mdx:333]:
```suggestion
  .queryDelimiter(",")
```

**Review by ikhoon — APPROVED** [review]
Thanks a lot, @ghkim3221!

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:175]: We missed adding `@UnstableApi`.

File: core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHostAnnotatedServiceBindingBuilder.java:183]: Ditto. We missed adding `@UnstableApi`.

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
Comment (ghkim3221) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java:176]: Nice catch!

**Comment by trustin** [comment]
🚀🚀🚀

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
Comment (ks-yim) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:724]: Late nit: Maybe we can have `Splitter` as a closure variable so that it doesn't get a new instance on every request?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:724]: @ghkim3221 What do you think?

**Pull request merged**
